javaslang/src/test/java/javaslang/collection/QuickSort.java

```diff
@@ -0,0 +1,25 @@
+/*     / \____  _    _  ____   ______  / \ ____  __    _______
+ *    /  /    \/ \  / \/    \ /  /\__\/  //    \/  \  //  /\__\   JΛVΛSLΛNG
+ *  _/  /  /\  \  \/  /  /\  \\__\\  \  //  /\  \ /\\/ \ /__\ \   Copyright 2014-2017 Javaslang, http://javaslang.io
+ * /___/\_/  \_/\____/\_/  \_/\__\/__/\__\_/  \_//  \__/\_____/   Licensed under the Apache License, Version 2.0
+ */
+package javaslang.collection;
+
+import org.junit.Test;
+
+import static javaslang.API.Vector;
+
+public class QuickSort {
+    @Test
+    public void shouldQuickSort() {
+        final Vector<Integer> values = Vector(3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5, 8, 9, 7, 9, 3, 2, 3, 8, 4, 6, 2, 6);
+        assert sort(values).equals(values.sorted());
+    }
+
+    /** Note: this example is only meant to show off, not to be used in reality: it can have quadratic performance and cause stack overflow */
+    private static Seq<Integer> sort(Seq<Integer> values) {
+        if (values.size() <= 1) return values;
+        else return values.tail().partition(v -> v <= values.head())
```

Comment:
Currently not. I plan to revise the patterns. The `List` Pattern will be deprecated. Instead we will have a `Cons` and a `Nil` pattern, directly encoded in javaslang.API.

You could define your own patterns, directly within Quicksort (I think / depends on the build order and pattern code generator - haven't tested it):

```java
public class Quicksort {

    ... // use Quicksort$Patterns.Cons here <-- not sure about the name of the class, too lazy to look it up

    @Patterns
    static class $ {
        @Unapply
        static <T> Tuple2<T, List<T>> Cons(List.Cons<T> cons) {
            return Tuple.of(cons.head(), cons.tail());
        }

        @Unapply
        static <T> Tuple0 Nil(List.Nil<T> nil) {
            return Tuple.empty();
        }
    }
}
```

For Seq there is a problem. Patterns are generated for ADTs. In particular an ADT is a _sum type_, which means that it is a disjoint union of several (two) implementations.

E.g. the ADT List is the union of List.Cons and List.Nil.

Seq is an interface. There is no unique union of types. But we could create a (head, tail) Pattern 'by hand' for the Traversable type:

```java
public class Quicksort {

    public static <T, _1 extends T, _2 extends Traversable<T>> Pattern2<Traversable<T>, _1, _2> HeadTail(Pattern<_1, ?> p1, Pattern<_2, ?> p2) {
        return new Pattern2<T, T1, T2>() {
                @SuppressWarnings("unchecked")
                @Override
                public Option<Tuple2<T1, T2>> apply(T obj) {
                    if (obj instanceof Traversable && ((Traversable) obj).isDefined()) {
                        final Traversable<T> traversable = (Traversable<T>) obj;
                        final Tuple2<U1, U2> unapplied = Tuple(traversable.head(), traversable.tail());
                        return unapplied.apply((u1, u2) ->
                                ((Pattern<U1, ?>) p1).apply(u1).flatMap(_1 ->
                                ((Pattern<U2, ?>) p2).apply(u2).map(_2 -> (Tuple2<T1, T2>) unapplied)
                        ));
                    } else {
                        return Option.none();
                    }
                }
            };
    }

}
```

Haven't tried it ... don't know if it compiles.